### PR TITLE
Add vault listings

### PR DESCRIPTION
**Pull request by tomwade**

### CODE REVIEW

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+pragma solidity ^0.8.0;
+
+import "./interface/INFTXVault.sol";
+import "./interface/INFTXVaultFactory.sol";
+import "./interface/INFTXVaultListing.sol";
+import "./interface/INFTXFeeDistributor.sol";
+import "./testing/IERC721.sol";
+import "./util/ReentrancyGuardUpgradeable.sol";
+
+import "hardhat/console.sol";
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:12]: Will drop console logging before merge

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+		uint nftId;
+		uint price;
+		bool active;
+		uint expiry;
+		address seller;
+	}
+
+    event ListingCreated();
+    event ListingUpdated();
+    event ListingFilled();
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:31]: Event attributes need populating

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+    // Flat store of all listings
+    Listing[] public listings;
+
+    // Listing vaults
+    mapping(address => uint[]) listingVaultIds;
+
+    uint public minFloorPrice;
+
+	INFTXVaultFactory vaultFactory;
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:44]: Tabs / spaces

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+    function __NFTXVaultListing_init(address _vaultFactory) public virtual initializer {
+    	vaultFactory = INFTXVaultFactory(_vaultFactory);
+    	minFloorPrice = 1200000000000000000;
+    }
+
+    // TODO: Need floor access
+    function setFloorPrice(uint _minFloorPrice) external {
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:52]: Need to add Floor access / onlyOwner

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+		uint vaultId;
+		address vaultAddress;
+		uint nftId;
+		uint price;
+		bool active;
+		uint expiry;
+		address seller;
+	}
+
+    event ListingCreated();
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:29]: `(uint256 listingId, uint256 nftId, address vault, uint256 price, uint256 expires)`

I don't believe we can do an event with all the data so would need to loop through each listing created and emit an event.

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    event ListingUpdated();
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:30]: `(uint256 listingId, uint256 price, uint256 expires)`

File: contracts/solidity/NFTXVaultListingUpgradeable.sol (hunk 2 above)
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:31]: `(uint256 listingId)`

**Review by 0xCaps — COMMENTED** [review]
Some event details...

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+        uint nftId;
+        uint amount;
+        uint price;
+        bool active;
+        uint expiry;
+        address seller;
+    }
+
+    // {ListingCreated} fired when a new listing is created
+    event ListingCreated(uint listingId, uint nftId, address vault, uint price, uint expires);
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:38]: Add amount

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+
+    // {ListingUpdated} fired when an existing listing is updated
+    event ListingUpdated(uint listingId, uint price, uint expires);
+
+    // {ListingFilled} fired when an existing listing is successfully filled
+    event ListingFilled(uint listingId);
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:44]: Add amount

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+        for (uint i; i < count;) {
+            // Don't let the user create a listing that expires in the past
+            require(expires[i] > block.timestamp, 'Listing already expired');
+
+            // Sanity check our pricing is above minimum
+            require(prices[i] >= minFloorPrice, 'Listing below floor price');
+
+            require(_senderOwnsNFT(msg.sender, vaults[i], nftIds[i], amounts[i]), 'Sender does not own NFT');
+
+            // ERC721 validation
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:112]: Move to internal function

Comment (tomwade) [reply contracts/solidity/NFTXVaultListingUpgradeable.sol:112]: Look at ERC165 to determine on-chain proof of implementation

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+            Listing memory existingListing = listings[listingId];
+
+            require(existingListing.active, 'Listing is not active');
+            require(existingListing.seller == msg.sender, 'Sender is not listing owner');
+
+            // Confirm the updated price is above minimum if changed
+            require(prices[i] >= minFloorPrice, 'Listing below floor price');
+
+            // Ensure our sender actually owners the NFT they are wanting to update
+            require(
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:171]: Close listing if user no longer owns the NFT

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+            address kitties = 0x06012c8cf97BEaD5deAe237070F9587f8E7A266d;
+            address punks = 0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB;
+
+            bytes memory data;
+            bool success;
+            bytes memory resultData;
+
+            if (asset == kitties) {
+                data = abi.encodeWithSignature("transferFrom(address,address,uint256)", from, to, nftId);
+            } else if (asset == punks) {
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:523]: Add punk test

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 12]
```diff
@@ -0,0 +1,679 @@
...
+            }
+
+            // If the NFT is no longer owned or is no longer approved, then we close the listing
+            // at the expense of the updater.
+            bool owned = _senderOwnsNFT(msg.sender, vault, nftId, amount);
+            bool approved = _senderApprovedNFT(msg.sender, vault, nftId);
+
+            if (!owned || !approved) {
+                return;
+            }
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:262]: Possibly remove this block

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 13]
```diff
@@ -0,0 +1,679 @@
...
+        address vault,
+        bytes32 listingId
+    ) internal {
+        // Confirm that our listing exists
+        require(_listingExists(listingId, false), 'Listing ID does not exist');
+
+        Listing721 storage existingListing = listings721[listingId];
+
+        // Confirm the buyer is not the seller
+        require(existingListing.seller != buyer, 'Buyer cannot be seller');
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:423]: Don't care

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 14]
```diff
@@ -0,0 +1,679 @@
...
+
+        // Send the seller tokens from the buyer
+        nftxVault.transferFrom(buyer, existingListing.seller, transferTokenAmount);
+
+        // Send NFT to buyer
+        _transfer(existingListing.seller, buyer, vault, nftId, 0);
+
+        // If we no longer have any amount in the listing, we can deactivate the
+        // listing by setting expiry time to 0.
+        _updateListing721(nftId, vault, listingId, existingListing.price, 0);
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:442]: Just `delete`

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 15]
```diff
@@ -0,0 +1,679 @@
...
+        INFTXVault nftxVault = INFTXVault(vault);
+
+        // Send the seller tokens from the buyer
+        nftxVault.transferFrom(buyer, existingListing.seller, purchaseAmount);
+
+        // Send NFT to buyer
+        _transfer(existingListing.seller, buyer, vault, nftId, amount);
+
+        // Reduce the amount available in our listing
+        existingListing.amount = existingListing.amount - amount;
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:494]: SafeMath sub

**Review by 0xCaps — CHANGES_REQUESTED** [review]

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 16]
```diff
@@ -0,0 +1,679 @@
...
+     * the listing has it's expiry timestamp into the past, then it will set the listing
+     * to inactive. If a `0` value is set for the expiry timestamp then the listing will
+     * also be deleted from our mapping.
+     *
+     * @param nftIds The IDs of the NFTs being filled
+     * @param vaults The addresses of the NFTX vaults
+     * @param amounts Number of 1155 tokens, should be 0 for 721s
+     */
+
+    function fillListings(
```
Comment (0xCaps) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:376]: Consider how we handle fees

**Comment by tomwade** [comment]
Pre-edits:
v1

Post-edits:
v2

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 17]
```diff
@@ -0,0 +1,859 @@
...
+    /**
+     * @notice Clears any cached external NFTX vault data,
+     * 
+     * @param vault The NFTX Vault asset address
+     */
+
+    function purgeNFTXCache(address vault) external onlyOwner {
+        nftxVault1155[vault] = 0;
+        nftxVaultAsset[vault] = address(0);
+        nftxVaultMintFee[vault] = 0;
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:856]: Drop mint fee caching

File: contracts/solidity/NFTXVaultListingUpgradeable.sol [hunk 18]
```diff
@@ -0,0 +1,859 @@
...
+        if (vaultFactory.excludedFromFees(seller)) {
+            return;
+        }
+
+        // Send the mint fee to the fee distributooooor
+        if (mintFee > 0) {
+            // Mint fees directly to the distributor and distribute.
+            address feeDistributor = vaultFactory.feeDistributor();
+            nftxVault.transferFrom(buyer, feeDistributor, mintFee);
+            INFTXFeeDistributor(feeDistributor).distribute(nftxVault.vaultId());
```
Comment (tomwade) [inline contracts/solidity/NFTXVaultListingUpgradeable.sol:678]: Can we drop this? Gas saves

**Pull request closed without merging**